Site Editor: Persistent List View

Copons opened the pull request:
## Description

Fixes #27395

Convert the List View dropdown into a persistent sidebar occupying the same interface skeleton slot as the Inserter (`secondarySidebar`).

To achieve this, a number of widespread changes have been made:

- Create a new `components/secondary-sidebar` folder containing both the Inserter and the List View file, making the Site Editor's `editor.js` easier to read.
- Create a new `listViewPanel` reducer (and relative action and selector) in the same vein as those for the inserter panel and navigation sidebar. These three reducers are interconnected: only one can be open at any time, so the action that opens one will automatically close the others.
- Create a new `listView` icon containing the List View icon, which at this point would have been hardcoded and repeated in 3 different files. Also replace those hardcoded icons with the new one.
- Replace the use of `BlockNavigationDropdown` in the Site Editor header with a custom implementation that would fill the `secondarySidebar` with the `BlockNavigationTree` component, displaying the entire blocks hierarchy of the content, as opposed to the dropdown, which only shows the currently selected nested hierarchy (e.g. if the site title block is selected, the dropdown would only show up to the header template part).
- Clicking on an item in the List View sidebar will select the block in the canvas (This is a pre-existing functionality of `BlockNavigationTree` itself).
- Reuse the same toggle keyboard shortcut for the post editor List View dropdown: <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> (letter o, not number 0; see also the [issue](https://github.com/WordPress/gutenberg/issues/27395#issuecomment-738732612)).
- Add a new `highlightBlocksOnHover` prop to the `BlockNavigation` component to support highlighting the corresponding block in the canvas on hover and focus.

### TODO

- Accessibility!
- Highlight the list view item when hovering the block in the canvas (candidate for a follow up: https://github.com/WordPress/gutenberg/issues/28699).
- Figure out a sane keyboard navigation (more on this below).

### Stuff to discuss

The keyboard navigation.
This is complicated, and any changes will affect the original `BlockNavigation` component itself.
What makes things awkward is that selecting a block in the `BlockNavigation` moves the focus to the canvas, and selecting blocks in the canvas will change the `BlockNavigation` selected item.
A common scenario would be:
1. Open the sidebar.
2. Move with the up/down keys: this only highlights the items with a border, not select them.
3. Hit enter: this will select the item (it becomes black), and move the focus to the block in the canvas (without any obvious visual clues).
4. Move with up/down: this will actually move between blocks, which in turns moves the selected item in the sidebar.
5. The user (me in this case) is now lost. The only way to return the focus to the sidebar is toggle it off and on again.

In other words: once the user selects an item in the sidebar, the vertical traversal changes.
Whereas one would expect that traversing the block list is just a matter of going up/down, traversing the canvas is a different matter.
Horizontally aligned blocks might want an horizontal movement instead, practically blocking the traversal.

## How has this been tested?

- Install and activate a FSE-enabled theme such as TT1-Blocks.
- Open the Site Editor.
- Click on the "List View" icon, or use the <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> shortcut.
- Make sure the List View sidebar pops up.
- Play around with it to see if it works (or not) as expected.
- Make sure only one of these sidebar can be open at any time: list view, inserter, navigation.

## Screenshots

(Ignore the unstyled canvas, it's an unrelated issue)

<img width="864" alt="Screenshot 2021-02-02 at 12 37 30" src="https://user-images.githubusercontent.com/2070010/106601379-797f2a00-6553-11eb-9e5e-9d500fe3673b.png">

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Copons committed:
Move the Inserter into its own component for readability

Copons committed:
Also move the styles

Copons committed:
Duplicate the inserter behaviour into a new secondary sidebar

Copons committed:
Use BlockNavigationTree in the new sidebar

Copons committed:
Create a new navigationMenu icon to be reused across the editors

Copons committed:
Style the block navigation sidebar

mtias commented on packages/icons/src/library/navigation-menu.js:
```diff
@@ -0,0 +1,12 @@
+ */
+import { SVG, Path } from '@wordpress/primitives';
+
+const navigationMenu = (
```
This one seems to be named "block-navigation" in figma:

![image](https://user-images.githubusercontent.com/548849/106611632-7e9aa400-6568-11eb-9f01-136065a0f700.png)

@jasmussen can you check what should be the desired name?

Copons replied:
Nice find!
I'll wait for a confirmation and update the icon name.

Copons committed:
Add a keyboard shortcut to toggle the block navigation sidebar

Copons committed:
Focus first element on mount

Copons commented:
I've opened this for review and added a few points of discussion to the PR description.

Copons committed:
Add highlight on hover/focus

Copons committed:
Simplify hover handling

jameskoster commented:
One nice thing to add would be an indication of the currently-selected-blocks children:

<img width="504" alt="Screenshot 2021-02-03 at 12 24 27" src="https://user-images.githubusercontent.com/846565/106746791-e7daef80-661a-11eb-9eff-9cb6dc6b44d5.png">

This can be a helpful visual cue for when a block has many innerblocks.

---

Also wondering if we need the padding around the items?

https://user-images.githubusercontent.com/846565/106748278-eb6f7600-661c-11eb-9a08-98baf86f0548.mp4

We might give the canvas more room by eliminating it.

On a similar note, the icons feel especially hefty now. I wonder if we might shrink them down a bit here, and tweak the spacing slightly? Here's 16px icons, and 8/12px padding on the buttons:

<img width="889" alt="Screenshot 2021-02-03 at 13 01 47" src="https://user-images.githubusercontent.com/846565/106750645-ff68a700-661f-11eb-8357-0f04678d5742.png">

Copons committed:
Rename the sidebar to "List View"

Copons committed:
Replace hardcoded store string

jameskoster commented:
I made some styling adjustments:

* Width now matches the Inspector
* Rows are full width
* Focus shadow

Additional css classes will be required to achieve this though:

<img src="https://user-images.githubusercontent.com/846565/106746791-e7daef80-661a-11eb-9eff-9cb6dc6b44d5.png">

jameskoster committed:
Styles

jameskoster committed:
radius

Copons commented:
@jameskoster I don't think we should have a fixed width: the `BlockNavigation` component renders a `GridTree` which internally uses a `table` element, which has a tendency of overflowing when its contents are larger than the available width.
In other words, this is what happens:

<img width="285" alt="Screenshot 2021-02-04 at 16 03 10" src="https://user-images.githubusercontent.com/2070010/106919956-82622e00-6702-11eb-84e4-6a66ab866118.png">

On the other hand, with an automatic width, if we have a very long block name, or a very deeply nested block, we might end up with an exceptionally large sidebar.

Given how the style changes need to be tested carefully, as they risk affecting the `BlockNavigation` component used elsewhere, I'd propose to move them to a follow-up.
None of them look like blockers to this PR to me, which works (or not! 😄 ) regardless of the items spanning edge-to-edge.

jameskoster commented:
I am open to moving general styling to a separate issue, but I think we should get the width sorted in this PR as that feels foundational to the overall usability of this feature.

I do not think the width should vary from page to page, or based on which blocks / nesting exists within the current document. Like you said, there's potential for a large variance which would feel quite disorientating to use.

So, options... I see only two at this point:

1. Simply hide/truncate the overflow. Not ideal, but there would need to be _a lot_ of nesting for this to cause a significant problem. Curiously this is what Figma [does](https://cloudup.com/i2JgfReMiZK) in their Layers panel.
2. [`overflow-x: auto;`](https://cloudup.com/iwe_fizO9Jq)

Any other ideas?

shaunandrews commented:
FWIW, we might want to update treeGrid to not use tables.

jasmussen commented:
> Simply hide/truncate the overflow. Not ideal, but there would need to be a lot of nesting for this to cause a significant problem. Curiously this is what Figma does in their Layers panel.
> overflow-x: auto;

I had to look at what Figma does. It truncates it, doesn't allow scroll, but does allow you to resize the width of the left sidebar:

<img width="344" alt="Screenshot 2021-02-05 at 09 21 51" src="https://user-images.githubusercontent.com/1204802/107008097-a883de80-6793-11eb-8dbc-70fbde5d9b42.png">

A 3r option is to do some smartness on truncation and the hiding of parents, so top level you see everything but truncated, but as you click downwards to select, you only ever see 2 or 3 or 4 parent layers up, whatever works. This could work as a sort of guided horizontal drilldown scrolling. 

But this is probably complicated and might not work well, and could always be added later, so for now I think you should just go with overflow: auto;

mtias commented:
I'd +1 moving the design discussion and reformulation to a separate issue. The implementation details are fairly tricky on these components so we'd want to avoid changing them too erratically until we are certain of the design details.

Copons committed:
Move the list item border radius and box shadow overrides to the persistent sidebar styles

Copons commented:
In 290264d I've moved the border radius and inset box shadow from the generic `BlockNavigation` component to `ListViewSidebar` only.
This way, the original list view dropdown will continue to have its original 2px radius and default box shadow.

---

Currently, the sidebar already has an horizontal scrollbar (not sure how, but I missed it yesterday), so I think this PR is good to go design-wise, pending a11y review.

The outstanding visual changes to be addressed in follow ups:
- Reduce icons size to 16px.
- Highlight children on parent item selection/hover (we should talk about it before starting dev).
- Figure out a way to handle wide lists to avoid both truncating items and the horizontal scrollbars.

Am I forgetting anything?

Copons committed:
Fix tests for new "List view" label

Copons committed:
Revert "Styles"

This reverts commit d4dd62d07e716b2f3c12930c263e0b2d519cb69f.

Copons committed:
Revert "Move the list item border radius and box shadow overrides to the persistent sidebar styles"

This reverts commit 47befcfc8afe92a86fb598f9248867ea67e39b2f.

Copons committed:
Revert "radius"

This reverts commit 7e57716d7801ed0a03b7ff77199a122e6c09f31c.

paaljoachim commented:
Another example from Figma.

![Screen Shot 2021-02-17 at 18 12 21](https://user-images.githubusercontent.com/5323259/108241060-b8b09c00-714b-11eb-8b04-e7811888a1ea.png)

Figma: Top parent is highlighted. (This PR) Nice to hear that selected parent and its children will be highlighted.
Figma: Each top menu item has an arrow so one can open or close the branch. Being able to open/close a branch makes it easier to focus on a specific area of the layout.

Copons commented:
Quick update after the discussion during the a11y + design joint office hours on Slack:

- The keyboard navigation works (roughly) as expected.
  - The List View sidebar can be toggled with <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> and navigated with the <kbd>up</kbd>/<kbd>down</kbd> arrows.
  - Hitting <kbd>enter</kbd> selects the block and moves the focus to the block in the canvas.
  - ⚠️ There is no easy way to move the focus back to the sidebar. It's possible to use <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to toggle it off and on again, or <kbd>tab</kbd> all the way back to it, but of course neither are ideal.
  Do we need another shortcut to go back to the sidebar at the selected block?

- When a block is selected, the corresponding List View item is selected as well.
  - ⚠️ With two "synchronized" selected elements on the screen, it's not clear where the focus is.
  - ⚠️ Especially confusing is the flow: "select block from the sidebar" -> "move focus to canvas" -> "press up/down to change block" -> "the selected item in the sidebar changes as well".
  - ⚠️ The list items need one more status to explicitly show where is the focus. These are just some suggestions to give the idea of what I mean:
    - Default: white background, no borders.
    - Hover and focus: grey background, no borders.
    - Selected and focused: black background, maybe with a `--wp-admin-theme-color` border.
    - Selected, but the block in the canvas has the focus: white background with a `--wp-admin-theme-color` border.

jameskoster commented:
> With two "synchronized" selected elements on the screen, it's not clear where the focus is.

> The list items need one more status to explicitly show where is the focus. These are just some suggestions to give the idea of what I mean:
> * Default: white background, no borders.
> * Hover and focus: grey background, no borders.
> * Selected and focused: black background, maybe with a --wp-admin-theme-color border.
> * Selected, but the block in the canvas has the focus: white background with a --wp-admin-theme-color border.

There are visual treatments applied to blocks on the canvas that we can potentially cross-pollinate to the list view. I believe these align to the status outlined above.

* Selected and focussed blocks have a 1.5px blue outline
* Selected but not focussed blocks have a 1px black outline
* Hovered blocks have a 1px blue outline

This could be a starting point for addressing this issue?

>  Especially confusing is the flow: "select block from the sidebar" -> "move focus to canvas" -> "press up/down to change block" -> "the selected item in the sidebar changes as well".

Can we elaborate on this? When a block is selected on the canvas, the list view should also be updated to reflect that. So what is described above seems fine? 🤔 

---

All that said... it was mentioned above that we should explore these design details in a separate issue. I think that's probably a good idea due to how widespread this component is.

paaljoachim commented:
I asked @Copons during the a11y and design chat to create a Gutenberg plugin version containing this PR making it easier for folks to test out. Here is the link https://github.com/WordPress/gutenberg/suites/2044639453/artifacts/41149125 (direct link to the zip file which contains the Gutenberg plugin zip file.) 

Instructions if you are on a Mac.
Download the above link.
Double click to unzip (as the Gutenberg plugin zip is inside of it will at the same time also be unzipped).
(Mac automatically unzips the plugin zip inside.)
Right click and choose to compress the unzipped Gutenberg plugin.
Upload to the plugin screen of your test site the now zipped up special version of the Gutenberg plugin.
Follow Copons instructions for testing seen in the initial post.

Copons commented:
> All that said... it was mentioned above that we should explore these design details in a separate issue. I think that's probably a good idea due to how widespread this component is.

True, but IMHO this is kind of a big deal. 😄 

> > Especially confusing is the flow: "select block from the sidebar" -> "move focus to canvas" -> "press up/down to change block" -> "the selected item in the sidebar changes as well".
> 
> Can we elaborate on this? When a block is selected on the canvas, the list view should also be updated to reflect that. So what is described above seems fine? 🤔

I've recorded a short video about the confusing flow I'm talking about:

https://user-images.githubusercontent.com/2070010/108354815-e811e800-71e1-11eb-915f-c822cc6c594c.mp4

This is what happens:

- Open the List View.
- Start navigating with the down arrow: the item is in focus state (blue border).
- Press enter: the item is now selected (black background), and the actual block as well (blue border).
- Keep navigating with the down arrow: both the list item and the canvas follow by changing the selection at the same time.
- **Q1**: Can you tell where is the focus?
  - The focus moved to the canvas, but (imho) the list item's black background is stronger than the block's blue border, so I'm (personally) tricked into thinking that the focus remained in the list.
- Keep navigating with the down arrow: at some point it will start skipping items in the List View.
- **Q2**: Can you tell what is happening?
  - I'm now navigating in the canvas, where the blocks exist both vertically and horizontally, and have many other navigation rules (or issues?), and pressing the down arrow won't necessarily move to the very next block.
  I might be aware of it if I knew I was in the canvas but, given the list item selection is moving along with my down arrow, I'm still thinking I'm navigating the list.


> There are visual treatments applied to blocks on the canvas that we can potentially cross-pollinate to the list view. I believe these align to the statuses outlined above.
> 
> * Selected and focussed blocks have a 1.5px blue outline
> * Selected but not focussed blocks have a 1px black outline
> * Hovered blocks have a 1px blue outline
> 
> This could be a starting point for addressing this issue?

Most definitely!
I'll give it a try and see if it helps with my confusing flow! 🙂

jasmussen commented:
Great thoughts, @copons. 

I think what you have here works really well, and there can potentially be enhancements to the focus question you mention: maybe the selected state in the canvas, when focus is on the sidebar, is only 1px thick? I.e. the "hover" state that Jay mentions here https://github.com/WordPress/gutenberg/pull/28637#issuecomment-780769485.

But it seems like the primary challenge you bring up is with the active state in the sidebar. The black total inversion is indeed pretty heavy. For alignments, we have a slightly more subtle active state:

<img width="585" alt="Screenshot 2021-02-19 at 11 30 23" src="https://user-images.githubusercontent.com/1204802/108493042-2bce2580-72a6-11eb-8e85-08badd974901.png">

That could be something to explore, perhaps combined with a gray border or background? What do you think @jameskoster ?

jameskoster commented:
As above I think it would be good to try mimicking the on-canvas status treatments in list view. In my head it works well, but until we can actually play around with it, I can't say for sure.

Whether that treatment translates effectively to other situations in which this component is used is another matter. I'm inclined to say no, but again perhaps it's something we have to try?

Copons committed:
Experimental: Try the same hover/focus/select colors for both blocks and list view items

Copons commented:
@jameskoster 78e1eea is an experimental commit where I've updated the colors to mirror the same hover/focus/select of blocks.
Feel free to peruse, and then I'll adjust the code appropriately, or drop it altogether. 🙂 

Here's a video for max convenience:

https://user-images.githubusercontent.com/2070010/108542037-1c1d0400-72db-11eb-93c6-08a8d38d30af.mp4

Copons commented on packages/block-editor/src/components/block-navigation/style.scss:
```diff
@@ -233,3 +233,27 @@
+			}
+		}
+	}
+}
```
NOTE: all these additions are currently **experimental**, and shouldn't be reviewed as final.

The idea is to try the new styles out and see if they fits the visual expectations.
Then, I'll move them where appropriate, or just remove them.

jameskoster commented:
Hmm, it's still a little confusing, huh? 

I remain optimistic though! I noticed a few instances where the styles were not applied as expected. Perhaps addressing the following could help:

* When I mouseover blocks in list view, I notice that blocks on the _canvas_ have the focus style (1.5px outline). They should instead have the hover style (1px outline), and only take on the focus styles if the block in the list view is clicked or the <kbd>return</kbd> key is pressed.
* Probably linked to the above, if you open list view with no prior selected block, the first block in the list has the hover style, and the corresponding block on the canvas has the focus style. I _think_ this should be the other way around? IE. Opening list view should shift focus to it.
* If you select a block on the canvas and then open list view, I think focus should move to that same block in list view. 

This is rather tricky to get right though, so I would encourage all thoughts and feedback.

mtias commented:
> When I mouseover blocks in list view, I notice that blocks on the canvas have the focus style (1.5px outline). They should instead have the hover style (1px outline), and only take on the focus styles if the block in the list view is clicked or the return key is pressed.

I'm thinking we should actually try to consolidate on the 1.5px like we are using here for the hover / focus effect: https://github.com/WordPress/gutenberg/pull/28598

Which has the same purpose as this one.

jasmussen commented:
> I'm thinking we should actually try to consolidate on the 1.5px like we are using here for the hover / focus effect

I'm coming around to this.

I've suggested 1px to ambiguate from focus in the past, because we should be very careful with blurring the lines between actual focus, and “block selected”, whenever it’s possible.

However in the name of simplicity, and because the behavior is tied to _hover_, to an extent it's no different than when you are hovering the block icon in the block toolbar to indicate the block boundaries. And it does keep things simpler.

jameskoster commented:
> Which has the same purpose as this one.

Ah I hadn't seen that change. I agree the purpose is the same, so the visual treatment should be as well 👍

The only issue here is that is-highlighted and is-selected styles are identical, which I worry undermines the overall heuristic. Perhaps that is something we can address elsewhere.

jameskoster commented:
If we consolidate hovered and highlighted to have a 1.5px outline, that means hovered, highlighted, _and_ selected+focussed states would all have the exact same style treatment. That could be a good point from which to explore a more holistic solution, but long term I don't know that it is acceptable? Perhaps I am missing something though, it feels like we've gone in a few circles here haha.

> the actual focus, not the visual effect

Shouldn't these always be aligned? :D 

> I think list and canvas should have some kind of synchronized scroll

They do already, but it only seems to be one-way. Clicking a block in list view scrolls the canvas accordingly.

Copons commented:
> If we consolidate hovered and highlighted to have a 1.5px outline, that means hovered, highlighted, _and_ selected+focussed states would all have the exact same style treatment. That could be a good point from which to explore a more holistic solution, but long term I don't know that it is acceptable? Perhaps I am missing something though, it feels like we've gone in a few circles here haha.

Possibly!
I think I'm losing grasp of all these tiny differences! 😅 

> > the actual focus, not the visual effect
> 
> Shouldn't these always be aligned? :D

They are, kinda!
In the latest styling changes, the list item's hover and focus have the same 1px outline.

![image](https://user-images.githubusercontent.com/2070010/108877672-5f3ae800-75f7-11eb-955a-41f7d1f52985.png)

 
> > I think list and canvas should have some kind of synchronized scroll
> 
> They do already, but it only seems to be one-way. Clicking a block in list view scrolls the canvas accordingly.

True!
Although, it would be also cool imho to scroll the canvas when navigating the list via keyboard.
Like, you press the down arrow in the list, and the block comes into view _before_ selecting it.

Copons committed:
Focus selected list item on mount

Copons committed:
Remove experimental style changes

Copons committed:
Add inline docs to the new focus hook

Copons committed:
Code cleanup

youknowriad commented on packages/edit-site/src/components/secondary-sidebar/list-view-sidebar.js:
```diff
@@ -0,0 +1,107 @@
+		}
+
+		const selectedItem = node.querySelector(
+			'.block-editor-block-navigation-leaf.is-selected'
```
instead of trying to find the element using classnames, can't we use a list of refs instead, since it seems it's all part of the same list view component?

Copons replied:
Actually... I think there's an even better way!

https://github.com/WordPress/gutenberg/blob/b3127331294c094ff115d1bf9b9d08439471b5d4/packages/block-editor/src/components/block-navigation/block.js#L82-L86

The code is already there, but it's coupled with the BlockNavigation's experimental features (e.g. drag and drop, a more menu for all the list items).
It just makes sense to reuse that for our purposes as well.

Copons committed:
Use the block navigation context to handle experimental persistent list view features

Copons commented:
A brief summary of the changes pushed today:

- Remove all experimental style changes from the List View. We'll follow up on them in a separate issue.

- Consolidate the `highlightBlockOnHover` and "focus on selected item" into a single feature, accessible by enabling the `__experimentalPersistentListViewFeatures` prop of the `BlockNavigation` component.
The new prop follows the same logic and inheritance of the existing `__experimentalFeatures`, which is kept and shared through the component context.
The new prop is a bit of a mouthful, but imho it makes it very clear how these features are currently only supposed to be used in one specific case.
The idea is that, if we feel good about them, we can (almost) just remove the flag and call it a day.

With this done, I think this PR is ready for a good old code review! ✨

jeyip commented:
## Testing
### Behaviors
- [x] Highlighting blocks in list view highlights the corresponding block in the canvas
- [x] Selecting blocks in list view selects the corresponding block in the canvas
- [x] Selecting blocks in the canvas selects the corresponding block in the list view
- [x] Only one of the list view, inserter, or navigation sidebar can be open at any given time
- [x] Ctrl + Alt + O hotkey toggles the list view open and closed 

### Browsers
- [ ] Edge
- [ ] Firefox
- [ ] Safari
- [x] Chrome

**Note:**
Will update this comment as a I test more behaviors and browsers

jeyip commented on packages/edit-site/src/components/secondary-sidebar/style.scss:
```diff
@@ -0,0 +1,49 @@
+	}
+}
+
+.edit-site-editor__list-view-panel {
```
`NB:`

Maybe we could consolidate some of the duplicated styles for `.edit-site-editor__inserter-panel-content`, `.edit-site-editor__list-view-panel-content`, `.edit-site-editor__inserter-panel`, and `.edit-site-editor__list-view-panel`?

jeyip commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -198,40 +185,12 @@ function Editor( { initialSettings } ) {
-															/>
-														</div>
-													</div>
-												) : null
```
Removing this `null` return value is what's causing the secondary sidebar from the interface skeleton to block the editor in mobile views (mentioned it [here](https://github.com/WordPress/gutenberg/pull/28637#issuecomment-785419490))

jeyip reviewed (commented):
`NB:`

~Do `list-view` and `block-navigator` refer to the same thing? If so, the mixed usage of terms seems confusing, and I wonder if it would be worth refactoring the names in the future.~

For others code reviewing, the block navigator and list view do seem to be different. From my understanding, the block navigator refers to all of the logic and UI associated with rendering the block navigation modal and the block navigation dropdown. 

## Modal
<img width="715" alt="Screen Shot 2021-02-24 at 4 26 57 PM" src="https://user-images.githubusercontent.com/5414230/109084889-7cd28500-76bd-11eb-834c-f0b947840278.png">

## Dropdown
<img width="1187" alt="Screen Shot 2021-02-24 at 6 24 24 PM" src="https://user-images.githubusercontent.com/5414230/109093651-8fed5100-76cd-11eb-8877-bcd691580d21.png">

I'm curious though -- is there a reason we don't follow the convention and call the list-view-sidebar the block-navigation-sidebar?

jeyip commented on packages/edit-site/src/store/reducer.js:
```diff
@@ -169,13 +170,36 @@ export function blockInserterPanel( state = false, action ) {
+ * @param {Object} state Current state.
+ * @param {Object} action Dispatched action.
+ */
+export function listViewPanel( state = false, action ) {
```
Can we add unit tests for this?

alexstine commented:
Here is some much over due accessibility feedback.

Tested with Firefox using NV Access on Windows 7.

The first issue I spotted was after opening the List View, pressing Escape does nothing. It should close the List View and place focus back on the trigger button.

The second issue is the keyboard shortcut is not always reliable. For example, if I open the List View with CTRL+Alt+O and then close the List View by selecting the List View button, the keyboard shortcut will not open the List View again.

The third issue I spotted was nothing is announced when the List View opens. It's like a void. Open List View and I expect to here something but nothing happens. Focus is put at the right place, but nothing indicates to me that the List View has opened. Sometimes I will here it announce "Block List Library", but it is not clear to me what a "Block List Library" is after I just selected a button that said "List View". Maybe it wouldn't hurt to try to describe more what List View button does.

Hope this helps.

Thanks.

Copons committed:
Fix secondary sidebar hiding the editor on small screens

InterfaceSkeleton's secondarySidebar prop always render something unless it's falsy,
so passing a component that renders null is not enough.

By handling the Inserter and List View sidebars visibility
from the parent component, I had to also move the
List View shortcut so that it's always available,
not just when the sidebar is visible.

Props to @jeyip for catching this!

Copons committed:
Consolidate duplicate styles

Copons commented:
@jeyip Thanks for the great catches! ✨ 

> It looks like the secondary sidebar from the interface skeleton is blocking the editor in mobile views

Wowee, I was aware of this but my brain still thought the Site Editor was broken on mobile... 🤦 
Anywho, the fix was a bit involved, because the `InterfaceSkeleton` renders the secondary sidebar even if it's an empty component. I had to move a few things around unfortunately, but nothing major — and I still think the Site Editor's `editor.js` is 100x more readable this way. 🙂 

> Not sure if this is expected or will be handled in a follow-up, but multi-selecting blocks in the canvas does not select corresponding blocks in the list view

Good one!
This happens because the `BlockNavigation` component uses `getSelectedBlockClientId` (singular) to highlight the selected item, whereas multiselection relies on `getSelectedBlockClientIds` (plural), which returns an array of all selected blocks, whether it's just one or many.
I gave a quick look at how the component handles the selection, and it seems a bit deep-rooted in several different logics (not to mention that the prop is passed down through almost the whole hierarchy).
I'll give it a try just in case, but I think it's a good fit for a follow up.

> Width of the list view appears to shrink when using the navigation sidebar to load other templates / parts in the editor

I think that's because the List View doesn't have a fixed width, but adjusts depending on the content.
The TT1 home is more nested and has a few blocks with long names, while Sample Page don't.

> I'm curious though -- is there a reason we don't follow the convention and call the list-view-sidebar the block-navigation-sidebar?

It was a design decision, mostly caused by the fact that it was already called in _many_ different names (list view, block navigation/navigator, outline), and everybody liked "List View" over the other options.

I wasn't aware of that button in the Navigation Block.
Don't have any strong opinions, but I think in that specific context "Block Navigator" is more appropriate than "List View".
@mtias do you have any additional insights?

> Can we add unit tests for this and the corresponding actions and reducers?

I was 100% sure I already did it! 😭

Copons committed:
Add useDialog hook to handle close on ESC

Copons committed:
Properly handle close on esc and focus return

Copons commented:
> Here is some much over due accessibility feedback.
> 
> Tested with Firefox using NV Access on Windows 7.
> 
> The first issue I spotted was after opening the List View, pressing Escape does nothing. It should close the List View and place focus back on the trigger button.
> 
> The second issue is the keyboard shortcut is not always reliable. For example, if I open the List View with CTRL+Alt+O and then close the List View by selecting the List View button, the keyboard shortcut will not open the List View again.
> 
> The third issue I spotted was nothing is announced when the List View opens. It's like a void. Open List View and I expect to here something but nothing happens. Focus is put at the right place, but nothing indicates to me that the List View has opened. Sometimes I will here it announce "Block List Library", but it is not clear to me what a "Block List Library" is after I just selected a button that said "List View". Maybe it wouldn't hurt to try to describe more what List View button does.
> 
> Hope this helps.
> 
> Thanks.

Hi @alexstine, again thanks for testing this PR!

> The first issue I spotted was after opening the List View, pressing Escape does nothing. It should close the List View and place focus back on the trigger button.

Nice catch! I didn't think of this at all!
I've replicated roughly the same behaviour of the Inserter sidebar.
Pressing <kbd>ESC</kbd> when the focus is inside the sidebar now closes the sidebar, returning the focus to the toggle button.
Although, I've noticed that opening via <kbd>ctrl</kbd>+<kbd>alt</kbd>+<kbd>O</kbd> shortcut and then closing with <kbd>ESC</kbd> doesn't guarantee the return focus.
I'm not entirely sure why, to be honest. It might be a glitch with the code handling the return focus (not part of this PR). 🤔 

> The second issue is the keyboard shortcut is not always reliable. For example, if I open the List View with CTRL+Alt+O and then close the List View by selecting the List View button, the keyboard shortcut will not open the List View again.

I can't reproduce this.
I can open and close the sidebar multiple times with the shortcut, or by mixing shortcut, mouse click, and <kbd>ESC</kbd>.

> The third issue I spotted was nothing is announced when the List View opens.

Not sure where "Block List Library" comes from, but I've added "List View" as ARIA label of the sidebar.
Do you think this would do the trick?
I'm testing with MacOS VoiceOver, which is probably not the greatest tool for a11y testing, and it sounds good to me?
Meaning, when I tab onto the toggle button it says "List View button", then when I open it, it reads the selected button (e.g. "Header button selected").

Copons committed:
Add tests for the new listViewPanel reducer

alexstine commented:
Everything is working now besides the keyboard shortcut. I have refreshed multiple times and cleared my browser cache. I will mention that I am using the built Gutenberg plugin from the build artifact and not manually building myself, but surely that's not the cause?

It won't work at all now. Pressing CTRL+ALT+O on Windows does nothing.

Thanks.

youknowriad commented on packages/edit-site/src/components/secondary-sidebar/list-view-sidebar.js:
```diff
@@ -28,12 +35,29 @@ export default function ListViewSidebar() {
+
+	const constrainedTabbingRef = useConstrainedTabbing();
 	const focusOnMountRef = useFocusOnMount( 'firstElement' );
+	const focusReturnRef = useFocusReturn();
```
Can't you just replace all these three hooks with useDialog instead?

Copons replied:
Not quite: `useDialog` also closes on click outside, which kinda defeats the "persistent" in "persistent list view". 😅 

(FYI, I did try it in 68606ba6372c6e12fae144073bf8b2e138404ae9, before noticing the click-outside behaviour)

youknowriad replied:
> Not quite: useDialog also closes on click outside

As I shared on the issue, this is the exact problem I tried to hightlight. I'm not sure removing that behavior is a good thing in terms of a11y.

youknowriad replied:
In other words, IMO, a persistent sidebar for me need a "checkbox"/"toggle" somewhere that says "show/hide the sidebar" but it shouldn't have a button that "opens the sidebar and moves focus to the sidebar" as well. It should have more a "setting" rather than an button that doesn't an "action" (open and navigate) as otherwise moving out would logically mean "close and navigate back"

youknowriad replied:
Here's a concrete problem of the current implementation:

 - Open the sidebar
 - Click outside
 - Click "Tab" repeatedly until you reach the sidebar, you'll get stuck in the sidebar because of the "constrained tabbing" behavior (which should only apply to dialogs) which means we created a "tab trap" which violates the WCAG  rules.

Copons replied:
TBH I think the current behaviour works well enough. 🤔 
The List View is openable via keyboard in 2 different ways (tabbing to the toggle button, or via shortcut), closable in 2 ways as well (ESC, or via shortcut), and can move focus to the canvas by selecting a list item.

This said, this stuff is very particular and I don't know all the context and discussions happened for the other sidebars, so I'm happy to disagree and update anyway.
Although, I've been working following designers mockups, and they might be more familiar with the established sidebar interactions? cc @jameskoster

youknowriad replied:
I know that's what the designers want, I'm just trying to illustrated that there's bug in the proposed behavior :). 

The only alternative I can think are:

Make the button in the top bar behave as a "setting" and not as an action. Which means don't focus the sidebar on mount. (Might make navigating to the sidebar a bit harder for screen readers users as it'd require a shortcut or using the region navigation but at least it keeps the keyboard navigation sane)

The other would would be to move the button to open the sidebar to the actual sidebar (top of it), you know just like the toggles we see in websites sometimes to show left menus. This means there's no need for focus management since the sidebar is shown right after the button. I'm sure designers can find a way to make this look good  :P  

Maybe there are other alternatives, but if we want it to be a persistent sidebar, we shouldn't be trying to make it work like a dialog (focus on mount and constrained tabbing), these things only work if you have (close on focus leave).

jameskoster replied:
> Make the button in the top bar behave as a "setting" and not as an action.

It's funny you mention that, because I happen to think this direction could be the way to go in the long term for various UI components. Toggles for things like list view and breadcrumb could potentially be added to the "View" menu (related: https://github.com/WordPress/gutenberg/pull/29151). Subsequently the existing toggles for top toolbar, spotlight mode, full-screen mode, etc, could conceivably join them there as well.

Rough mockup for this context:

![block list](https://user-images.githubusercontent.com/846565/109196293-6c8cd980-7793-11eb-8365-b984f2b50822.gif)

> The other would would be to move the button to open the sidebar to the actual sidebar

I'm struggling to picture how this would work. If the button is inside the sidebar, and the sidebar is hidden, how do you click the button? :D

Copons replied:
@youknowriad apologies, I missed the [comment](https://github.com/WordPress/gutenberg/pull/28637#discussion_r582988035) where you outlined the issue! Must have popped up while I was writing my answer.

I didn't notice how the constrained tabbing affected the sidebar, but you are totally right about it.
I've removed the constrained tabbing leaving only these behaviours:
- focus on mount (selected element if any, or first element)
- close on ESC
- focus return

The tab is not trapped anymore, although the keyboard navigation is still confusing.
The `BlockNavigation` component renders a tree, which is navigable with the arrows instead of tab.
If you are focusing a list item, tab moves you to the canvas, selecting a block, which in turns highlights the corresponding list item. At that point it's very hard to understand if you are focusing the canvas or the list.

Though, this confusion happens even when clicking around with the mouse, as the selected list item is very heavy and in-your-face, compared to the selected block outline.
We have discussed about it for a while, and decided to move it to a follow up dedicated to the style of the sidebar.

I think that if you are cool with the fact that the tab is not trapped anymore, we can move forward with this for both the style and a decision about how to appropriately toggle the sidebar.

IMHO this PR is really valuable in itself even with its rough edges, and it's limited to the Site Editor so it won't cause nasty regressions in production features.
In other words, I'm keen to merge this (provided the code is good, of course), and iterate to fine tune the UX.

Addison-Stavlo replied:
> The other would would be to move the button to open the sidebar to the actual sidebar

> > I'm struggling to picture how this would work. If the button is inside the sidebar, and the sidebar is hidden, how do you click the button? :D

I think this idea is that the toggle button should be above the sidebar (similar to where the list view button currently exists outside of this PR) and more 'connected' to it.  Since it would be a 'setting' and not directing focus into the sidebar, it would be important for the next 'tab' from the toggle button to lead into the sidebar if it is open.

youknowriad replied:
> I'm struggling to picture how this would work. If the button is inside the sidebar, and the sidebar is hidden, how do you click the button? :D

I was suggesting here that when it's "closed", it's not "entirely closed" but instead the "button" is the only thing in the sidebar (and the sidebar doesn't have any background or borders, so we only see the button)

That said, moving to "View" is actually very good and is what makes the most sense to me. 

Now, about the current behavior (after dropping the constrained tabbing), I think I'm ok with it personally, though it's arguable whether other screen reader users might think.

Copons committed:
Remove the constrained tabbing

Addison-Stavlo reviewed (commented):
Incoming wall of text 😅  - mostly behavioral observations and me getting myself confused.  I don't really expect us to come up with solutions for 1 or 2 in this PR, but perhaps we could handle item 3.

> Move with up/down: this will actually move between blocks, which in turns moves the selected item in the sidebar.
The user (me in this case) is now lost. The only way to return the focus to the sidebar is toggle it off and on again.

100% - In testing this out I was immediately lost and  **very** confused as to what was going on with how navigation worked in various circumstances.  It turns out a lot of that confusion came from uncertainty (or ignorance) between when I am focused in the list view, when I am focused in the editor, and the navigation differences between them.

#### 1 - I was immediately confused (the editor got focus in 'select' mode when I hit 'enter' and 'tab' from the toggle button).
The first big confusion I experienced was when I navigated to the toggle button, hit 'enter' to open the sidebar, and then hit 'tab'.  Up/down arrowkeys wouldnt take me into the indented levels, i would have to hit the right arrowkey to access the 'buttons' for example. Then in that level up/down would not let me exit until hitting the left arrowkey.  I felt locked in and it felt very awkward:
![unexpected-what](https://user-images.githubusercontent.com/28742426/109231951-6566cb00-7795-11eb-910d-54a59ece40f2.gif)

Now, I was mistaking the black highlighted list view button as having the focus.  It turns out that first 'tab' took focus into the editor, and since nothing was previously selected the editor defaulted to select mode (explaining the oddness between levels and left/right arrowkeys to navigate between indentation levels) and selected the first top level group.  While the behavior is explainable/understandable, this was definitely a source of confusion for a bit.  No suggestions for how to make that better in the scope of this PR, but figured I would point it out.

#### 2 - Getting focus back into list view requires closing/reopening sidebar.
Now when we have used this properly (unlike 1 above) and select a block through the list-view, we have moved focus into the editor and up/down navigation differs in the editor than it does in list view (up/down keys for example may skip blocks depending on layout).  While this makes sense, as a keyboard user I am wanting to get back into the sidebar so I can navigate with a consistent up/down behavior.  To do this I need to `ctrl+alt+o`  twice to close/reopen the panel to move focus back into it.  It would be nice to be able to move focus back into the item i was on in the sidebar without having to close/reopen it, although shft+tab would still navigate the editor and I don't think a dedicated hotkey just to move focus back to it makes sense. 🤔   Id think maybe `ctrl+alt+o` could move focus back to it if it is open and not focused, and close it if it is focused... but then that seems like it could get confusing as well.  So once again, no immediate solutions that I can propose there.

#### 3. Selecting a block in list view doesn't focus in editor if block is already selected in editor state.

After I close/re-open the sidebar with my block selected in the editor, it opens with focus on that item in the sidebar.  Great!  However, then selecting that item again with 'enter' does nothing. Since the block is already technically selected in the editor state, clicking the button to select it does not 'do anything' for focus and focus remains trapped in the sidebar.  I would then need to select another block in the list view to move focus back into the editor:

![bad-enter](https://user-images.githubusercontent.com/28742426/109234964-be852d80-779a-11eb-965e-ba1bcbba1769.gif)

Notice in the gif above I start with the cursor in button 'two'.  I hotkey close/open the sidebar to move focus back into it to navigate consistently.  I decide I actually do want to edit button 'two' and try to hit 'enter' to select it.  This doesn't work.  Eventually, I select button 'one' and then arrow key over to button two once focus is back in the editor.  It would be nice if hitting 'enter' on an item in the list view would still move focus to the block even if it was already selected.

#### 4. Using the block inserter closes the list view.

For a 'persistent list view' to be truly persistent, it should probably persist beyond the user opening and selecting an item from the block inserter sidebar:

![not-persistent](https://user-images.githubusercontent.com/28742426/109236395-87644b80-779d-11eb-80e3-e0b48f486596.gif)

The interaction between these two sidebars is definitely a bit odd and hard to define.  But inserting blocks is a fairly common action and if the list view disappears after doing so, its not really fitting the need for a 'persistent' navigation tool. 🤔

jeyip commented:
> I gave a quick look at how the component handles the selection, and it seems a bit deep-rooted in several different logics (not to mention that the prop is passed down through almost the whole hierarchy).
I'll give it a try just in case, but I think it's a good fit for a follow up.

A follow-up sounds good to me 👍 

> I think that's because the List View doesn't have a fixed width, but adjusts depending on the content.
The TT1 home is more nested and has a few blocks with long names, while Sample Page don't.

I wonder if having a min-width is worthwhile for the sake of UI consistency 🤔 

> It was a design decision, mostly caused by the fact that it was already called in many different names (list view, block navigation/navigator, outline), and everybody liked "List View" over the other options.
>
> I wasn't aware of that button in the Navigation Block.
> Don't have any strong opinions, but I think in that specific context "Block Navigator" is more appropriate than "List View".

Oh it sounds like depending on what we decide, it may be worthwhile to have an issue to come back one day and consolidate the naming. (Doesn't have to be done in this PR)

> I was 100% sure I already did it! 😭

https://media.giphy.com/media/7OVSXLwZQm2xhBewpk/giphy.gif

Copons committed:
Make the List View as wide as the Inserter

Copons commented:
@Addison-Stavlo thanks for the in depth review! 🙇‍♂️ 

> 1 - I was immediately confused (the editor got focus in 'select' mode when I hit 'enter' and 'tab' from the toggle button).

Yep, this is imho the biggest pain point of this PR.
We have done a few backs and forths with design to figure out if there were any quick wins, but no luck.
The plan is to address it in a follow up, where we can discuss the styling requirements without all the technical burdens.

> 2 - Getting focus back into list view requires closing/reopening sidebar.

This is super tricky, and perfect for a follow up imho.

> 3. Selecting a block in list view doesn't focus in editor if block is already selected in editor state.

Blurgh, I'm not exactly sure what happens here.
My first guess was that maybe a selecting a selected block doesn't move the focus, but I've tried clearing the selection before selecting, and still the focus remains in the list.
This definitely requires more investigation, and at this point I have no clues. 😓 

> 4. Using the block inserter closes the list view.

Good point, but I'd rather not get stuck in semantics. 😄 
"Persistent" is more a project name than the actual feature (no mention of any persistence in the UI), so... 🤷 

There was a suggestion from design at some point, to have Inserter and List View to share the same sidebar _at the same time_, with both filling up half of the sidebar (or a resizable portion, whatever).

It's a discussion we'll need to have, but only after we see that the "persistent-ish" List View is actually helpful in practice.

---

> I wonder if having a min-width is worthwhile for the sake of UI consistency 🤔

@jeyip ✅

youknowriad reviewed (commented):
Should this be site editor specific? What would it take to do it in the post editor? Does it impact the block navigation of the post editor in its current form?

Copons commented:
> Should this be site editor specific? What would it take to do it in the post editor? Does it impact the block navigation of the post editor in its current form?

@youknowriad The idea is to try it out in the Site Editor, to avoid wreaking havoc to the majority of the users, and because the Site Editor is arguably where the List View is more needed.
If we see that it's a good improvement over the dropdown, we'll proceed to move it to the Post Editor as well.

Technically, we would need to copypaste the `listViewPanel` state over to `core/edit-post`, and the `ListViewSidebar` component to the Post Editor's interface.

All the changes to the common `BlockNavigation` component are behind an `__experimentalPersistentListViewFeatures` flag, so currently there should be no changes for any other usages.

youknowriad commented:
I just hope that it's something that we'll be follow-up on, we had a lot of situations where things are tried and then left as is. Consolidation is important to ensure the quality of the project long term.

Copons commented:
> It won't work at all now. Pressing CTRL+ALT+O on Windows does nothing.

@alexstine (and @Addison-Stavlo who was testing the same): it just occurred to me that on Win the shortcut is <kbd>shift</kbd><kbd>alt</kbd><kbd>o</kbd> (with shift instead of ctrl).

See the `access` modifier defined here: https://github.com/WordPress/gutenberg/blob/c993a862483510e8fab9e949201a8ce89cb7c54a/packages/keycodes/src/index.js#L141

Addison-Stavlo commented:
Thanks for checking out my concerns @Copons!

> It's a discussion we'll need to have, but only after we see that the "persistent-ish" List View is actually helpful in practice.

I think this accurately describes most of what I have brought up and none of that needs to be a blocker here considering we are limited the scope of this change to the site editor as we test it out and get a feel for it.  Il see if I can figure out anything more regarding item 3 (it would definitely be nice to fix early) and will dig into the code side of things a bit more today.

Copons committed:
Fix typo

Addison-Stavlo commented on packages/block-editor/src/components/block-navigation/block.js:
```diff
@@ -63,38 +62,62 @@ export default function BlockNavigationBlock( {
 		}
-	}, [ withExperimentalFeatures, isSelected ] );
+	}, [
+		withExperimentalFeatures,
+		withExperimentalPersistentListViewFeatures,
+		isSelected,
```
This dep list is why trying `await clearSelectedBlock` before `selectBlock` was not solving issue 3.  If I make this an empty array and then try that solution, it starts working!  

I guess since `isSelected` will toggle from true -> false -> true in that situation, this `.focus` fires as the last event and focus remains in sidebar as opposed to moving to that selected block.  I admit that I have no idea why this isn't an issue normally, since it behaves that way I would expect this to draw focus back to the sidebar anytime I try to have it select a block (whether it is already selected or not). Either way, it definitely seems to be what is stopping the clear/reselect approach for working as expected.

I don't think there is any reason for this dep array not to be blank?  As far as I can tell the only time we would want to focus an item in the sidebar would be on mount.  But having `isSelected` as a dependency means `.focus()` will be called on this sidebar item when its corresponding block becomes the selected block, which seems quite problematic.

Addison-Stavlo commented on packages/edit-site/src/components/secondary-sidebar/list-view-sidebar.js:
```diff
@@ -0,0 +1,79 @@
+			selectedBlockClientId: getSelectedBlockClientId(),
+		};
+	} );
+	const { selectBlock } = useDispatch( blockEditorStore );
```
```suggestion
	const { selectBlock, clearSelectedBlock } = useDispatch( blockEditorStore );
	const selectEditorBlock = async ( clientId ) => {
		await clearSelectedBlock();
		selectBlock( clientId );
	};
```

Addison-Stavlo commented on packages/edit-site/src/components/secondary-sidebar/list-view-sidebar.js:
```diff
@@ -0,0 +1,79 @@
+			>
+				<BlockNavigationTree
+					blocks={ rootBlocks }
+					selectBlock={ selectBlock }
```
```suggestion
					selectBlock={ selectEditorBlock }
```

Addison-Stavlo reviewed (commented):
The rest of the code looks good to me!

Below are code change suggestions that fix the focus issue when selecting an already selected block.

Copons committed:
Fix focus handling

Props @Addison-Stavlo ✨

Copons commented:
Thanks for the suggestion @Addison-Stavlo!
I got stuck looking across the entire codebase, when the solution was right under my nose.

Pushed your change, and it seems the focus works as expected.
- Open List View with no selected block: focus on first list item.
- Open List View with a selected block: focus on corresponding list item.
- Click on a list item, both selected or otherwise: focus on the corresponding block.

Addison-Stavlo reviewed (approved):
Awesome!  That works well for me and everything else looks good.
As noted we can follow up on the other design and UX/UI uncertainties, but this seems in a good place to start testing out the feature!

Thanks for all your work on this @Copons !

gziolo commented:
I noticed the following confusing behavior. When you use arrow keys in List View and press `space`, then focus moves to the editor. However, you still might have an impression that the arrow up/down lets you navigate between items, but it skips some blocks because you are moving focus between blocks in the editor.

Copons commented:
> I noticed the following confusing behavior. When you use arrow keys in List View and press `space`, then focus moves to the editor. However, you still might have an impression that the arrow up/down lets you navigate between items, but it skips some blocks because you are moving focus between blocks in the editor.
> 
>  Screen.Recording.2021-03-03.at.17.03.02.mov

@gziolo Yes, we are (painfully) aware of it: https://github.com/WordPress/gutenberg/issues/29467 😄 

We talked at lenght about it in this PR too (at this point all buried in the hidden comments), we even went back and forth with some visual experiment, and eventually decided to move the discussion to a follow up, to avoid stalling this PR.

Designers are aware of it too, and are thinking of a nice solution as we speak. 👍
